### PR TITLE
Use official k8s watcher to watch for k8s events

### LINKED ISSUE #6252: Rethink k8s watcher implementation
In Cilium 1.3 we have introduced a wrapper(1) around k8s so it would stop re-adding the objects every time cilium was resyncing against kube-apiserver.

Unfortunately, this requires a local copy of those objects so we can compare new objects against old objects and verify if they actually need to perform an update. As we are also using kubernetes watcher, it also keeps a local copy of those objects so we end up having 2 copies of the same objects.

I took a quick look in `k8s.io/client-go/tools/cache/reflector.go` and unfortunately it looks like this won't be easy as the kubernetes watcher requires a copy of the data, in the form of a `store`. We could check if it is possible to re-implement our version of the kubernetes watcher that wouldn't require to have a `store` or, since we have all objects locally in (1), we could implement the `store` interface in that wrapper.

### PR DESCRIPTION
**Pull request by aanm**
Long story short: don't reinvent the wheel. Read per commit. ~Last commit can be seen as a bug fix, I can create a backport commit for that bug fix if anyone wants it.~

Fixes: #6252

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7301)

### CODE REVIEW

**Review by joestringer — COMMENTED** [review]
Janitors are requested for review here, but it looks like the agent/k8s teams could provide much better input. Acknowledging the PR but not reviewing at this stage.

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

File: daemon/k8s_watcher.go [hunk 1]
```diff
@@ -337,157 +337,194 @@ func (d *Daemon) EnableK8sWatcher(reSyncPeriod time.Duration) error {
 		return fmt.Errorf("Unable to create cilium network policy client: %s", err)
 	}
 
+	serKNPs := serializer.NewFunctionQueue(1024)
```
Comment (tgraf) [inline daemon/k8s_watcher.go]: Let's make the 1024 configurable via an option.

File: pkg/k8s/factory_functions_test.go [hunk 2]
```diff
@@ -25,85 +25,10 @@ import (
 	. "gopkg.in/check.v1"
 	core_v1 "k8s.io/api/core/v1"
 	"k8s.io/api/extensions/v1beta1"
-	"k8s.io/api/networking/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
-func (s *K8sSuite) Test_missingK8sNetworkPolicyV1(c *C) {
```
Comment (tgraf) [inline pkg/k8s/factory_functions_test.go:33]: I assume this should be folded into the previous commit?

File: daemon/k8s_watcher.go [hunk 3]
```diff
@@ -367,6 +367,10 @@ func (d *Daemon) EnableK8sWatcher() error {
 					metrics.EventTSK8s.SetToCurrentTime()
 					if oldK8sNP := k8s.CopyObjToV1NetworkPolicy(oldObj); oldK8sNP != nil {
 						if newK8sNP := k8s.CopyObjToV1NetworkPolicy(newObj); newK8sNP != nil {
+							if k8s.EqualV1NetworkPolicy(oldK8sNP, newK8sNP) {
+								return
```
Comment (tgraf) [inline daemon/k8s_watcher.go:372]: I think we need to cover this with a metric somehow to understand how much we are ignoring. Ideally we bump a metric for each event received here and have a label type = { invalid | equal | valid }

I have some ideas here. Happy to work on this after we merge this PR.

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by aanm** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Pull request merged**